### DISCUSSION TITLE
BOOTSEL via Reset button

### DISCUSSION THREAD

**Discussion by fapplin** [earlephilhower/arduino-pico#892] [Q&A] [answered]
I have added a button that is connected to GND and RUN. It resets just fine. I would like to be able to add the bootsel capability via hitting the reset button twice. The SDK has the following info.

pico_bootsel_via_double_reset - When the 'pico_bootsel_via_double_reset' library is linked, a function is injected before main() which will detect when the system has been reset twice in quick succession, and enter the USB ROM bootloader (BOOTSEL mode) when this happens. This allows a double tap of a reset button on a development board to be used to enter the ROM bootloader, provided this library is always linked. 

Is this capability in this core? Is there anything I need to do code-wise?

Thanks,

Frank

**Comment by earlephilhower** [comment, marked as answer]
Can you try #893?  Add `rp2040.enableDoubleResetBootloader()` to your setup and it should link in the SDK code to check for a double reset tap.

**Reply by fapplin** [reply]
Is it exposed? I loaded my sketch in the 2.0 version of the IDE and the autocomplete doesn't show that method and I get the error:  error: 'class RP2040' has no member named 'enableDoubleResetBootloader. 

Frank

**Reply by earlephilhower** [reply]
You need to use the PR under a `git` copy.    It will hilight and compile properly then.  It's just that I don't have any boards w/RESET buttons so I can't test myself...

**Reply by fapplin** [reply]
I guess I'm not sure quite how to do that. I followed these instructions:

# Installing via GIT
To install via GIT (for latest and greatest versions):
````
mkdir -p ~/Arduino/hardware/pico
git clone https://github.com/earlephilhower/arduino-pico.git ~/Arduino/hardware/pico/rp2040
cd ~/Arduino/hardware/pico/rp2040
git submodule update --init
cd pico-sdk
git submodule update --init
cd ../tools
python3 ./get.py
`````
I've always installed via the board manager. How do I get the arduino IDE to use this?

**Reply by earlephilhower** [reply]
You'll need to do the above and then run `git checkout d1` to get the patched version (or you can use the official GH pull flow which is a little complicated).  Then the IDE should show something like "RP2040...(in sketchbook)" that you can select.

**Reply by fapplin** [reply]
OK. We're making progress. I do see it now in the autocomplete, but i get the following errors:
````
/home/frank/.arduino15/packages/rp2040/tools/pqt-gcc/1.4.0-c-0196c06/bin/../lib/gcc/arm-none-eabi/10.3.0/../../../../arm-none-eabi/bin/ld: /tmp/arduino-sketch-FBD47D57FD9EDD0924CED44F459E96B4/sketch/Improv_V02.ino.cpp.o: in function `_Z6setup1v':
/home/frank/Arduino/Improv_V02/Improv_V02.ino:481: undefined reference to `_ZN9WiFiClass5beginEPKcS1_'
/home/frank/.arduino15/packages/rp2040/tools/pqt-gcc/1.4.0-c-0196c06/bin/../lib/gcc/arm-none-eabi/10.3.0/../../../../arm-none-eabi/bin/ld: /tmp/arduino-sketch-FBD47D57FD9EDD0924CED44F459E96B4/sketch/Improv_V02.ino.cpp.o: in function `setup':
/home/frank/Arduino/Improv_V02/Improv_V02.ino:416: undefined reference to `_ZN6RP204027enableDoubleResetBootloaderEv'
````

Am I missing something or have the statement in the wrong place? (and, thank you again for your help)

**Reply by earlephilhower** [reply]
The first bit is that you re trying to build a WiFi project on a Pico, not PicoW.  Check the menus to make sure you have the right board.

And the second one is saying that I forgot to check in the new file, d'oh!  Please do  `git pull` to update and retry.

**Reply by fapplin** [reply]
OK. Got it compiled and loaded, but I get no result. The Reset using a button on the RUN and GND works fine. I'm expecting that if I hit the button twice (rapidly) that it will go into the mode where the Pico shows up as a USB drive. Isn't that what the function is suppose to do? That's not what is happening.

**Reply by fapplin** [reply]
And, just as aside. I am running 2 cores. So, I tried it in Setup and Setup1 (and both - just in case) and that didn't help. Is it function that needs to be called pre-setup?

**Reply by earlephilhower** [reply]
No, one time is enough.  The function doesn't actually do anything at runtime, it just makes sure the constructor is linked in which does the real work.

I think the issue is that there is OTA in this core which is probably overwriting their little marker in RAM.  So I need to set a special spot for that marker and make sure OTA ignores it...

**Reply by fapplin** [reply]
OK. It's not any kind of rush. I'm just thinking ahead on how people will update the code on this device I've created easily. I will have a Reset button just in case something locks up. So, I thought I'd use it for double duty. You don't know how grateful I am (and I know other are) for your core!!

**Reply by earlephilhower** [reply]
Do a `git pull` and try again.  I've just hacked together an old PC reset switch and some hookup wire and tested it to verify it's working on my own system now.

**Reply by fapplin** [reply]
My LittleFS upload seems to have gone away in the 2.0 IDE tools menu. So, I'm getting an error No drive to deploy. Do I need to go back to the old IDE?

**Reply by fapplin** [reply]
OK. Got it working in the 1.8.19 IDE!!!!! Reset and BOOTSEL on double tap!!!

**Reply by earlephilhower** [reply]
> My LittleFS upload seems to have gone away in the 2.0 IDE tools menu. So, I'm getting an error No drive to deploy. Do I need to go back to the old IDE?

Yup, there is no plugin support at all for IDE 2.0.  See https://github.com/arduino/arduino-ide/issues/58

**Reply by fapplin** [reply]
That sucks. The 2.0 IDE is almost a real IDE finally! Thank you again for your help!
